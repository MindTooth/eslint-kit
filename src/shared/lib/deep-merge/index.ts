/* eslint-disable @typescript-eslint/no-explicit-any */
import { DeepMerge, PathsOf } from './types'

function isObject(value: any): value is Record<string, any> {
  return typeof value === 'object' && value !== null
}

export enum Merge {
  Shallow,
  Deep,
  Override,
}

type AllPaths<A, B> = PathsOf<A> | PathsOf<B> | null

// eslint-disable-next-line sonarjs/cognitive-complexity
export function deepMerge<A extends any, B extends any>(
  a: A,
  b: B,
  strategy: (path: AllPaths<A, B>) => Merge = () => Merge.Deep,
  basePath: string | null = null,
  root = true,
  shallow = false
): DeepMerge<A, B> {
  if (Array.isArray(a) && Array.isArray(b)) {
    return [...a, ...b] as DeepMerge<A, B>
  }

  if (isObject(a) && isObject(b)) {
    if (shallow) {
      return { ...a, ...b } as DeepMerge<A, B>
    }

    if (root) {
      if (strategy(null) === Merge.Shallow) {
        return deepMerge(a, b, strategy as any, null, false, true) as DeepMerge<
          A,
          B
        >
      }

      if (strategy(null) === Merge.Override) {
        return b as DeepMerge<A, B>
      }
    }

    const keys = [...Object.keys(a), ...Object.keys(b)]
    const final: any = {}

    for (const key of keys) {
      const path = basePath ? [basePath, key].join('.') : key

      if (strategy(path as any) === Merge.Shallow) {
        final[key] = deepMerge(
          a[key],
          b[key],
          strategy as any,
          path,
          false,
          true
        )
      } else if (strategy(path as any) === Merge.Override) {
        final[key] = b[key]
      } else {
        final[key] = deepMerge(a[key], b[key], strategy as any, path, false)
      }
    }

    return final
  }

  if (typeof b !== 'undefined') {
    return b as DeepMerge<A, B>
  }

  return a as DeepMerge<A, B>
}
