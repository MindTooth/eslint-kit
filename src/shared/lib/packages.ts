import { parse } from 'semver'

type Dependencies = Record<string, string>

export interface PackageJson {
  version: string
  dependencies?: Dependencies
  devDependencies?: Dependencies
}

function mergeDependencies(packageJson: PackageJson) {
  const { dependencies = {}, devDependencies = {} } = packageJson
  return { ...dependencies, ...devDependencies }
}

export function getPackageSemver(name: string, packageJson: PackageJson) {
  const allDependencies = mergeDependencies(packageJson)
  const version = allDependencies[name]
  if (!version) return null
  const parsed = parse(version)
  if (!parsed) return null
  return parsed
}
