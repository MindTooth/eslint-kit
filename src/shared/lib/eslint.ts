import { Linter } from 'eslint'
import { deepMerge, Strategy } from './deep-merge'

function parseExtension(pattern: string): string | null {
  if (!pattern.startsWith('*.')) return null
  return pattern.slice(2)
}

function splitOverrides(
  overrides: Linter.ConfigOverride[]
): Linter.ConfigOverride[] {
  const result: Linter.ConfigOverride[] = []

  for (const { files, ...rest } of overrides) {
    const patterns = Array.isArray(files) ? files : [files]
    for (const pattern of patterns) {
      result.push({ files: [pattern], ...rest })
    }
  }

  return result
}

export function overrideOverrides(
  overrides: Linter.ConfigOverride[]
): Linter.ConfigOverride[] {
  const result: Linter.ConfigOverride[] = []

  const reversedOverrides = overrides.slice().reverse()

  const starRules = new Set<string>()
  const usedRules = new Map<string, Set<string>>()

  for (const { files, rules = {}, ...rest } of reversedOverrides) {
    const [pattern] = Array.isArray(files) ? files : [files]

    const extension = parseExtension(pattern)
    if (!extension) continue

    if (!usedRules.has(extension)) {
      usedRules.set(extension, new Set<string>())
    }

    const extensionUsedRules = usedRules.get(extension) ?? new Set()
    if (!usedRules.has(extension)) usedRules.set(extension, extensionUsedRules)

    const overridedRules: Linter.RulesRecord = {}

    for (const [ruleName, ruleEntry] of Object.entries(rules)) {
      if (!ruleEntry) continue
      if (extensionUsedRules.has(ruleName)) continue
      if (starRules.has(ruleName)) continue
      overridedRules[ruleName] = ruleEntry
      extensionUsedRules.add(ruleName)
      if (extension === '*') starRules.add(ruleName)
    }

    result.push({ files, rules: overridedRules, ...rest })
  }

  return result
}

export function mergeConfigs(configs: Linter.Config[]): Linter.Config {
  const config = configs.reduce((final, current) => {
    return deepMerge(final, current, (path) => {
      if (path === 'rules') return Strategy.Shallow
      return Strategy.Deep
    })
  })

  const splittedOverrides = splitOverrides(config.overrides ?? [])
  const overrides = overrideOverrides(splittedOverrides)

  return { ...config, overrides }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type OnlyCollections<T> = T extends any[] ? T : T extends object ? T : never

function applier<
  T extends 'rules' | 'settings' | 'extends' | 'parserOptions' | 'overrides'
>() {
  return (
    condition: unknown,
    entity: OnlyCollections<Linter.Config[T]>
  ): OnlyCollections<Linter.Config[T]> => {
    const fallback = Array.isArray(entity) ? [] : {}
    if (!condition) return fallback as OnlyCollections<Linter.Config[T]>
    return entity
  }
}

export const conditional = {
  rules: applier<'rules'>(),
  settings: applier<'settings'>(),
  extends: applier<'extends'>(),
  parserOptions: applier<'parserOptions'>(),
  overrides: applier<'overrides'>(),
}
