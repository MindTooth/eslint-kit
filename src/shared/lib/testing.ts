import fs from 'fs'
import path from 'path'
import { ESLint, Linter } from 'eslint'

export interface Options {
  config: unknown
  basePath: string
  files: string[]
  extension: 'js' | 'ts' | 'tsx'
}

export async function testConfig({
  config,
  basePath,
  files,
  extension,
}: Options) {
  const cli = new ESLint({
    baseConfig: config as Linter.Config<Linter.RulesRecord>,
    useEslintrc: false,
    cwd: process.cwd(),
    ignore: false,
  })

  for (const file of files) {
    const filePath = path.resolve(basePath, `./tests/${file}.${extension}`)
    const code = fs.readFileSync(filePath).toString()
    const isTS = ['ts', 'tsx'].includes(extension)
    const result = await cli.lintText(code, {
      filePath: isTS ? filePath : undefined,
    })
    expect(result).toMatchSnapshot(`${file}`)
  }
}
