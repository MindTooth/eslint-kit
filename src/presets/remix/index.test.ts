import { testConfig } from '../../shared/lib/testing'
import { react } from '../react'
import { remix } from './index'

describe('[Presets] Remix', () => {
  it('should match snapshots', async () => {
    await testConfig({
      presets: [react(), remix()],
      dirname: __dirname,
      files: ['basic-clear'],
      extension: 'js',
    })
  })
})
