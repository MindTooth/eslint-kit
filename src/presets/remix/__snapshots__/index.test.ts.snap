// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Presets] Remix should match snapshots: basic-clear 1`] = `
Array [
  Object {
    "errorCount": 0,
    "fatalErrorCount": 0,
    "filePath": "<PROJECT_ROOT>/src/presets/remix/tests/basic-clear.js",
    "fixableErrorCount": 0,
    "fixableWarningCount": 0,
    "messages": Array [],
    "suppressedMessages": Array [],
    "usedDeprecatedRules": Array [],
    "warningCount": 0,
  },
]
`;
